docs: add 'Flutter plugin' reference (#5755)
canonical/snapcraft: docs/reference/plugins/flutter_plugin.rst

.. _reference-flutter-plugin:

Flutter plugin
==============

The Flutter plugin simplifies the building of applications that employ the `Flutter

> <https://flutter.dev/>`_ framework.

Reviewer: ```suggestion
<https://flutter.dev/>`_ app framework.
```

Or maybe "UI"? A nitpick, but it might be nice to be more qualitative in what it is.

Keywords
--------

This plugin provides the following unique keys for core22 and newer snaps.

flutter-channel
~~~~~~~~~~~~~~~
**Type**: ``stable``, ``master``, or ``beta``

> **Default**: ``stable``

Reviewer: I noticed you fixed the linebreaks in these. Not an issue, but we should eventually go back and tweak the others here and the ones in Craft Parts. Similar to how we need to correct the "Keyword" heading names.

Author: Both of those problems have been bugging me for a little while. I might just open a quick PR so this doesn't get forgotten.

The default flutter channel to use for the build.

flutter-target
~~~~~~~~~~~~~~
**Type**: string

**Default**: ``lib/main.dart``

> The flutter target to build.

Reviewer: ```suggestion
The path to the app's entrypoint ``.dart`` file.
```

Dependencies
------------

For core22 and newer snaps, this plugin installs ``clang``, ``curl``, ``git``,

> ``cmake``, ``ninja-build``, and ``unzip``.

Reviewer: Might want to make this an unordered list?

Author: https://github.com/canonical/snapcraft/pull/5755/commits/715b9b6cbdee4d26ac3bbb8f7533e6c759b2f01d

How it works
------------

During the build step, the plugin performs the following actions:

#. Clone `Flutter's GitHub repository <https://github.com/flutter/flutter>`_, using the
   channel declared by the ``flutter-channel`` key. If a channel isn't specified, the
   ``stable`` channel is cloned.
#. Populate Flutter's cache of binary artifacts with ``flutter precache``.
#. Download any dependencies listed in the Flutter project's ``pubspec.yaml`` with
   ``flutter pub get``.

> #. Build the target declared by ``flutter-target`` with ``flutter build``. If no target
>    is specified, ``lib/main.dart`` is built.

Reviewer: ```suggestion
#. Build the app with ``flutter build``.
```

I don't think we need to reiterate the entrypoint file here. This threw me off when reading.

Author: https://github.com/canonical/snapcraft/pull/5755/commits/715b9b6cbdee4d26ac3bbb8f7533e6c759b2f01d

.. _reference-flutter-plugin:

Flutter plugin
==============

> The Flutter plugin builds `Flutter <https://flutter.dev/>`_-based parts

Reviewer 2: The majority of snaps that have a part that uses the flutter plugin are flutter applications. Perhaps this could be reframed like around the app rather than the part itself?

```suggestion
The Flutter plugin simplifies building applications that use the `Flutter <https://flutter.dev/>`_ framework.
```

I don't think we need much more detail beyond this. At least I can't think of anything else to mention.

Keywords
--------

This plugin provides the following unique keys for core22 and core24 snaps.

flutter-channel
~~~~~~~~~~~~~~~
**Type**: ``stable``, ``master``, or ``beta``

**Default**: ``stable``

The default flutter channel to use for the build.

flutter-target
~~~~~~~~~~~~~~
**Type**: string

**Default**: ``lib/main.dart``

The flutter target to build.

Dependencies
------------

For core22 and core24 snaps, this plugin installs ``clang``, ``curl``, ``git``,
``cmake``, ``ninja-build``, and ``unzip``.

How it works
------------

During the build step, the plugin performs the following actions:

#. Clone `Flutter's GitHub repository <https://github.com/flutter/flutter>`_, using the
   channel declared by the ``flutter-channel`` key. If a channel isn't specified, the
   ``stable`` channel is cloned.
#. Remove ``flutter-distro/engine/src/.gn`` from the part's build directory so that

>    subsequent commands run successfully.

Author: Oh, nice! When I checked the issue, I saw that it auto-closed and overlooked [this comment](https://github.com/flutter/flutter/issues/163308#issuecomment-2676366371).

https://github.com/canonical/snapcraft/pull/5755/commits/2bc698be475662f8a80eefab187e4dd277650bca

#. Populate Flutter's cache of binary artifacts with ``flutter precache``.
#. Download any dependencies listed in the Flutter project's ``pubspec.yaml`` with
   ``flutter pub get``.
#. Build the target declared by ``flutter-target`` with ``flutter build``. If no target

>    is specified, ``lib/main.dart`` is built.

Reviewer 2: If this was autotools, I'd appreciate seeing `configure`, `make`, and `make install` commands here, as it's less cognitive load to see a familiar command than reading the definition of the command and trying to guess which command it's describing.

I imagine listing the flutter commands are equally helpful for flutter devs.
